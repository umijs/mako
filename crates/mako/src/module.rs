use std::collections::HashSet;
use std::fmt::{Debug, Formatter};
use std::path::PathBuf;
use std::sync::Arc;

use anyhow::{anyhow, Result};
use bitflags::bitflags;
use pathdiff::diff_paths;
use serde::Serialize;
use swc_core::common::{Span, DUMMY_SP};
use swc_core::ecma::ast::{
    BlockStmt, ExportSpecifier, FnExpr, Function, ImportDecl, ImportSpecifier, Module as SwcModule,
    ModuleExportName, NamedExport,
};
use swc_core::ecma::utils::quote_ident;

use crate::ast::css_ast::CssAst;
use crate::ast::file::{win_path, File};
use crate::ast::js_ast::JsAst;
use crate::build::analyze_deps::AnalyzeDepsResult;
use crate::compiler::Context;
use crate::config::ModuleIdStrategy;
use crate::resolve::ResolverResource;

pub type Dependencies = HashSet<Dependency>;

#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct Dependency {
    pub source: String,
    pub resolve_as: Option<String>,
    pub resolve_type: ResolveType,
    pub order: usize,
    pub span: Option<Span>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ModuleSystem {
    CommonJS,
    ESModule,
    Custom,
}

bitflags! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct ResolveTypeFlags: u16 {
        const Sync  = 1;
        const Async = 1<<2;
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct ImportType: u16 {
        const Default = 1;
        const Named = 1<<2;
        const Namespace = 1<<3;
        const SideEffect = 1<<4 ;
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct NamedExportType: u16 {
        const Named = 1;
        const Default = 1<<2;
        const Namespace = 1<<3;
    }
}

impl From<&ResolveType> for ResolveTypeFlags {
    fn from(value: &ResolveType) -> Self {
        match value {
            ResolveType::DynamicImport(_) | ResolveType::Worker(_) => Self::Async,
            _ => Self::Sync,
        }
    }
}

impl From<&ImportDecl> for ImportType {
    fn from(decl: &ImportDecl) -> Self {
        if decl.specifiers.is_empty() {
            ImportType::SideEffect
        } else {
            let mut import_type = ImportType::empty();
            for specifier in &decl.specifiers {
                match specifier {
                    ImportSpecifier::Named(_) => {
                        import_type |= ImportType::Named;
                    }
                    ImportSpecifier::Default(_) => {
                        import_type |= ImportType::Default;
                    }
                    ImportSpecifier::Namespace(_) => {
                        import_type |= ImportType::Namespace;
                    }
                }
            }
            import_type
        }
    }
}

impl From<&NamedExportType> for ImportType {
    fn from(value: &NamedExportType) -> Self {
        let mut res = Self::empty();
        value.iter().for_each(|b| match b {
            NamedExportType::Default => res.insert(Self::Default),
            NamedExportType::Named => res.insert(Self::Named),
            NamedExportType::Namespace => res.insert(Self::Namespace),
            _ => {}
        });
        res
    }
}

impl From<&NamedExport> for NamedExportType {
    fn from(decl: &NamedExport) -> Self {
        let mut res = Self::empty();

        decl.specifiers
            .iter()
            .for_each(|specifier| match specifier {
                ExportSpecifier::Namespace(_) => res.insert(Self::Namespace),
                ExportSpecifier::Default(_) => res.insert(Self::Default),
                ExportSpecifier::Named(named) => {
                    if let ModuleExportName::Ident(orig) = &named.orig
                        && orig.sym.eq("default")
                    {
                        res.insert(Self::Default);
                    } else {
                        res.insert(Self::Named);
                    }
                }
            });

        res
    }
}

#[derive(Eq, Hash, PartialEq, Serialize, Debug, Clone, Default)]
pub struct ImportOptions {
    pub chunk_name: Option<String>,
    pub ignore: bool,
}

impl ImportOptions {
    pub fn get_chunk_name(&self) -> &Option<String> {
        &self.chunk_name
    }
}

#[derive(Eq, Hash, PartialEq, Serialize, Debug, Clone)]
pub enum ResolveType {
    Import(ImportType),
    ExportNamed(NamedExportType),
    ExportAll,
    Require,
    DynamicImport(ImportOptions),
    Css,
    Worker(ImportOptions),
}

impl ResolveType {
    pub fn same_enum(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Import(_), Self::Import(_)) => true,
            (_, _) => self == other,
        }
    }
}

impl ResolveType {
    pub fn is_esm(&self) -> bool {
        self.is_sync_esm() || self.is_dynamic_esm()
    }

    pub fn is_sync_esm(&self) -> bool {
        matches!(
            self,
            ResolveType::Import(_) | ResolveType::ExportNamed(_) | ResolveType::ExportAll
        )
    }

    pub fn is_dynamic_esm(&self) -> bool {
        matches!(self, ResolveType::DynamicImport(_))
    }
}

#[derive(Debug, Clone)]
pub struct ModuleInfo {
    pub ast: ModuleAst,
    pub file: File,
    pub deps: AnalyzeDepsResult,
    pub external: Option<String>,
    pub raw: String,
    pub raw_hash: u64,
    /// Modules with top-level-await
    pub top_level_await: bool,
    /// The top-level-await module must be an async module, in addition, for example, wasm is also an async module
    /// The purpose of distinguishing top_level_await and is_async is to adapt to runtime_async
    pub is_async: bool,
    pub is_ignored: bool,
    pub resolved_resource: Option<ResolverResource>,
    /// The transformed source map chain of this module
    pub source_map_chain: Vec<Vec<u8>>,
    pub module_system: ModuleSystem,
}

impl Default for ModuleInfo {
    fn default() -> Self {
        Self {
            module_system: ModuleSystem::CommonJS,
            ast: ModuleAst::None,
            file: Default::default(),
            deps: Default::default(),
            external: None,
            raw: "".to_string(),
            raw_hash: 0,
            top_level_await: false,
            is_async: false,
            resolved_resource: None,
            source_map_chain: vec![],
            is_ignored: false,
        }
    }
}

fn md5_hash(source_str: &str, lens: usize) -> String {
    format!("{:x}", md5::compute(source_str))
        .chars()
        .take(lens)
        .collect::<String>()
}

pub fn generate_module_id(origin_module_id: &str, context: &Arc<Context>) -> String {
    match context.config.module_id_strategy {
        ModuleIdStrategy::Hashed => md5_hash(origin_module_id, 8),
        ModuleIdStrategy::Named => {
            // readable ids for debugging usage
            let absolute_path = PathBuf::from(origin_module_id);
            let relative_path = diff_paths(&absolute_path, &context.root).unwrap_or(absolute_path);
            win_path(relative_path.to_str().unwrap())
        }
        ModuleIdStrategy::Numeric => {
            let numeric_ids_map = context.numeric_ids_map.read().unwrap();
            if let Some(numeric_id) = numeric_ids_map.get(origin_module_id) {
                numeric_id.to_string()
            } else {
                md5_hash(origin_module_id, 8)
            }
        }
    }
}

pub fn relative_to_root(module_path: &String, root: &PathBuf) -> String {
    let absolute_path = PathBuf::from(module_path);
    let relative_path = diff_paths(&absolute_path, root).unwrap_or(absolute_path);
    // diff_paths result always starts with ".."/"." or not
    if relative_path.starts_with("..") || relative_path.starts_with(".") {
        relative_path.to_string_lossy().to_string()
    } else {
        PathBuf::from(".")
            .join(relative_path)
            .to_string_lossy()
            .to_string()
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct ModuleId {
    pub id: String,
}

impl Ord for ModuleId {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.id.cmp(&other.id)
    }
}

impl PartialOrd for ModuleId {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl ModuleId {
    // we use absolute path as module id now
    pub fn new(id: String) -> Self {
        Self { id }
    }

    pub fn generate(&self, context: &Arc<Context>) -> String {
        // TODO: 如果是 Hashed 的话，stats 拿不到原始的 chunk_id
        generate_module_id(&self.id, context)
    }

    pub fn from_path(path_buf: PathBuf) -> Self {
        Self {
            id: path_buf.to_string_lossy().to_string(),
        }
    }

    // FIXME: 这里暂时直接通过 module_id 转换为 path，后续如果改了逻辑要记得改
    pub fn to_path(&self) -> PathBuf {
        PathBuf::from(self.id.clone())
    }
}

impl From<String> for ModuleId {
    fn from(id: String) -> Self {
        Self { id }
    }
}

impl From<&str> for ModuleId {
    fn from(id: &str) -> Self {
        Self { id: id.to_string() }
    }
}

impl From<PathBuf> for ModuleId {
    fn from(path: PathBuf) -> Self {
        Self {
            id: path.to_string_lossy().to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum ModuleAst {
    Script(JsAst),
    Css(CssAst),
    None,
}

impl ModuleAst {
    #[allow(dead_code)]
    pub fn as_script(&self) -> Option<&JsAst> {
        match self {
            ModuleAst::Script(ast) => Some(ast),
            _ => None,
        }
    }

    pub fn script_mut(&mut self) -> Option<&mut JsAst> {
        match self {
            ModuleAst::Script(ast) => Some(ast),
            _ => None,
        }
    }

    pub fn as_script_ast(&self) -> &SwcModule {
        if let Self::Script(script) = self {
            &script.ast
        } else {
            panic!("ModuleAst is not Script")
        }
    }

    pub fn as_script_ast_mut(&mut self) -> &mut SwcModule {
        if let Self::Script(script) = self {
            &mut script.ast
        } else {
            panic!("ModuleAst is not Script")
        }
    }

    pub fn as_css_mut(&mut self) -> &mut CssAst {
        if let Self::Css(css) = self {
            css
        } else {
            panic!("ModuleAst is not Css")
        }
    }

    pub fn as_script_mut(&mut self) -> &mut JsAst {
        if let Self::Script(script) = self {
            script
        } else {
            panic!("ModuleAst is not Css")
        }
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum ModuleType {
    Script,
    Css,
    Raw,
    PlaceHolder,
}

#[derive(Clone)]
pub struct Module {
    pub id: ModuleId,
    pub is_entry: bool,
    pub info: Option<ModuleInfo>,
    pub side_effects: bool,
}

impl Module {
    pub fn new(id: ModuleId, is_entry: bool, info: Option<ModuleInfo>) -> Self {
        Self {
            id,
            is_entry,
            info,
            side_effects: is_entry,
        }
    }

    pub fn as_script(&self) -> Option<&JsAst> {
        self.info.as_ref().and_then(|i| i.ast.as_script())
    }

    pub fn as_mut_script(&mut self) -> Option<&mut JsAst> {
        self.info.as_mut().and_then(|i| i.ast.script_mut())
    }

    pub fn set_info(&mut self, info: Option<ModuleInfo>) {
        self.info = info;
    }

    pub fn is_external(&self) -> bool {
        self.info
            .as_ref()
            .map_or(false, |info| info.external.is_some())
    }

    pub fn is_placeholder(&self) -> bool {
        self.get_module_type() == ModuleType::PlaceHolder
    }

    pub fn get_module_type(&self) -> ModuleType {
        self.info
            .as_ref()
            .map_or(ModuleType::PlaceHolder, |info| match info.ast {
                ModuleAst::Script(_) => ModuleType::Script,
                ModuleAst::Css(_) => ModuleType::Css,
                ModuleAst::None => ModuleType::Raw,
            })
    }

    pub fn get_module_size(&self) -> usize {
        self.info
            .as_ref()
            .map_or(0, |info| info.raw.as_bytes().len())
    }

    // wrap module stmt into a function
    // eg:
    // function(module, exports, require) {
    //   module stmt..
    // }
    pub fn to_module_fn_expr(&self) -> Result<FnExpr> {
        match &self.info.as_ref().unwrap().ast {
            ModuleAst::Script(script) => {
                let mut stmts = Vec::new();

                for n in script.ast.body.iter() {
                    match n.as_stmt() {
                        None => {
                            return Err(anyhow!(
                                "Error: not a stmt found in {:?}, ast: {:?}",
                                self.id.id,
                                n,
                            ));
                        }
                        Some(stmt) => {
                            stmts.push(stmt.clone());
                        }
                    }
                }

                let func = Function {
                    span: DUMMY_SP,
                    ctxt: Default::default(),
                    params: vec![
                        quote_ident!("module").into(),
                        quote_ident!("exports").into(),
                        quote_ident!("__mako_require__").into(),
                    ],
                    decorators: vec![],
                    body: Some(BlockStmt {
                        span: DUMMY_SP,
                        ctxt: Default::default(),
                        stmts,
                    }),
                    is_generator: false,
                    is_async: false,
                    type_params: None,
                    return_type: None,
                };
                Ok(FnExpr {
                    ident: None,
                    function: func.into(),
                })
            }
            // TODO: css modules will be removed in the future
            ModuleAst::Css(_) => Ok(empty_module_fn_expr()),
            ModuleAst::None => Err(anyhow!("ModuleAst::None({}) cannot concert", self.id.id)),
        }
    }
}

impl Debug for Module {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "id={}({:?})", self.id.id, self.get_module_type())
    }
}

fn empty_module_fn_expr() -> FnExpr {
    let func = Function {
        ctxt: Default::default(),
        span: DUMMY_SP,
        params: vec![
            quote_ident!("module").into(),
            quote_ident!("exports").into(),
            quote_ident!("__mako_require__").into(),
        ],
        decorators: vec![],
        body: Some(BlockStmt {
            ctxt: Default::default(),
            span: DUMMY_SP,
            stmts: vec![],
        }),
        is_generator: false,
        is_async: false,
        type_params: None,
        return_type: None,
    };
    FnExpr {
        ident: None,
        function: func.into(),
    }
}
