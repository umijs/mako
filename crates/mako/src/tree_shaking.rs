use std::collections::HashMap;

use swc_ecma_visit::VisitMutWith;
use tracing::debug;

use crate::compiler::Compiler;
use crate::module::ModuleId;
use crate::tree_shaking_module::{ModuleSystem, TreeShakingModule};
use crate::unused_statement_marker::UnusedStatementMarker;

impl Compiler {
    pub fn tree_shaking(&self) {
        // 拓扑排序
        let (entry_modules, sorted_modules, cycle_modules) = self.make_toposort();
        debug!("entry_modules: {:?}", &entry_modules);
        debug!("cycle_modules: {:?}", &cycle_modules);

        // 入口模块设置为副作用模块
        self.markup_entry_modules_as_side_effects(entry_modules);

        // 循环依赖模块设置为副作用
        self.markup_cycle_modules_as_side_effects(cycle_modules);

        let (tree_shaking_module_ids, mut tree_shaking_module_map) =
            self.create_tree_shaking_module_map(sorted_modules);

        let tree_shaking_module_map = &mut tree_shaking_module_map;

        let mut modules_to_remove = vec![];
        for tree_shaking_module_id in tree_shaking_module_ids {
            let tree_shaking_module = tree_shaking_module_map
                .get_mut(&tree_shaking_module_id)
                .unwrap();

            debug!(
                "tree_shaking module: [{}]{}",
                tree_shaking_module.side_effects, &tree_shaking_module_id.id,
            );
            debug!(
                "    - used_exports: {:?}",
                &tree_shaking_module.used_exports
            );

            if matches!(tree_shaking_module.module_system, ModuleSystem::ESModule) {
                // 包含副作用的模块
                if tree_shaking_module.side_effects {
                    {
                        {
                            let imports = tree_shaking_module.imports();
                            let exports = tree_shaking_module.exports();

                            // 分析使用情况
                            for import in imports {
                                debug!("    - import: {:?}", &import);
                                self.analyze_import_statement(
                                    tree_shaking_module_map,
                                    &tree_shaking_module_id,
                                    import,
                                );
                            }

                            for export in exports {
                                debug!("    - export: {:?}", &export);
                                self.analyze_export_statement(
                                    tree_shaking_module_map,
                                    &tree_shaking_module_id,
                                    export,
                                );
                            }
                        };
                    }
                } else {
                    // 不包含副作用的模块，执行 tree shaking
                    let tree_shaking_module = tree_shaking_module_map
                        .get_mut(&tree_shaking_module_id)
                        .unwrap();

                    let mut module_graph = self.context.module_graph.write().unwrap();
                    let module = module_graph
                        .get_module_mut(&tree_shaking_module.id)
                        .unwrap();
                    let ast = module.info.as_mut().unwrap().ast.as_script_mut();

                    // 通过 tree_shaking_module 进行无用的标记
                    let mut comments = self.context.meta.script.output_comments.write().unwrap();
                    let mut marker = UnusedStatementMarker::new(tree_shaking_module, &mut comments);
                    ast.visit_mut_with(&mut marker);
                    drop(module_graph);
                    drop(comments);

                    // 当前模块没有被使用到的导出，删除当前模块
                    if tree_shaking_module.used_exports.is_empty() {
                        modules_to_remove.push(tree_shaking_module_id.clone());
                        continue;
                    }

                    {
                        let imports = tree_shaking_module.imports();
                        let exports = tree_shaking_module.exports();

                        // 分析使用情况
                        for import in imports {
                            debug!("    - import: {:?}", &import);
                            self.analyze_import_statement(
                                tree_shaking_module_map,
                                &tree_shaking_module_id,
                                import,
                            );
                        }

                        for export in exports {
                            debug!("    - export: {:?}", &export);
                            self.analyze_export_statement(
                                tree_shaking_module_map,
                                &tree_shaking_module_id,
                                export,
                            );
                        }
                    };
                }
            }
        }

        self.cleanup_no_used_export_module(modules_to_remove);
    }

    fn make_toposort(&self) -> (Vec<ModuleId>, Vec<ModuleId>, Vec<Vec<ModuleId>>) {
        let module_graph = self.context.module_graph.read().unwrap();
        let entry_modules = module_graph.get_entry_modules();
        let (sorted_modules, cycle_modules) = module_graph.toposort();
        (entry_modules, sorted_modules, cycle_modules)
    }

    fn markup_entry_modules_as_side_effects(&self, entry_modules: Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for entry_module_id in entry_modules {
            let module = module_graph.get_module_mut(&entry_module_id).unwrap();
            module.side_effects = true;
        }
    }

    fn markup_cycle_modules_as_side_effects(&self, cycle_modules: Vec<Vec<ModuleId>>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for cycle_module in cycle_modules {
            for module_id in cycle_module {
                let module = module_graph.get_module_mut(&module_id).unwrap();
                module.side_effects = true;
            }
        }
    }

    #[allow(dead_code)]
    fn cleanup_no_used_export_module(&self, modules_to_remove: Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        debug!("modules_to_remove: {:?}", &modules_to_remove);
        for module_id in modules_to_remove {
            module_graph.mark_missing_module(&module_id, &self.context);
        }
    }

    fn create_tree_shaking_module_map(
        &self,
        sorted_modules: Vec<crate::module::ModuleId>,
    ) -> (Vec<ModuleId>, HashMap<ModuleId, TreeShakingModule>) {
        let mut tree_shaking_module_ids = vec![];
        let mut tree_shaking_module_map = HashMap::new();
        for module_id in sorted_modules {
            let mut dependencies_of_module: Vec<ModuleId> = vec![];
            {
                let module_graph_r = self.context.module_graph.read().unwrap();
                dependencies_of_module.extend(
                    module_graph_r
                        .get_dependencies(&module_id)
                        .iter()
                        .map(|f| f.0.clone())
                        .collect::<Vec<_>>(),
                );
            }

            let mut module_graph = self.context.module_graph.write().unwrap();
            let module = module_graph.get_module_mut(&module_id).unwrap();

            // external 模块，设置为副作用模块
            if module.is_external() {
                module.side_effects = true;
            }

            if !module.get_module_type().is_script() {
                // 非 js 模块，以及他们的依赖，都设置为副作用模块
                for dep_module_id in dependencies_of_module {
                    let dep_module = module_graph.get_module_mut(&dep_module_id).unwrap();
                    dep_module.side_effects = true;
                }
            }

            let module = module_graph.get_module(&module_id).unwrap();
            let tree_shaking_module = TreeShakingModule::new(module);
            tree_shaking_module_ids.push(tree_shaking_module.id.clone());
            tree_shaking_module_map.insert(tree_shaking_module.id.clone(), tree_shaking_module);
        }
        (tree_shaking_module_ids, tree_shaking_module_map)
    }
}

#[cfg(test)]
mod tests {
    use crate::assert_display_snapshot;
    use crate::test_helper::{read_dist_file, setup_compiler};

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking() {
        let compiler = setup_compiler("test/build/tree-shaking", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_reexport", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_named_reexport", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_namespace() {
        let compiler = setup_compiler("test/build/tree-shaking_export_namespace", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_export() {
        let compiler = setup_compiler("test/build/tree-shaking_named_export", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_fn() {
        let compiler = setup_compiler("test/build/tree-shaking_fn", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_side_effect() {
        let compiler = setup_compiler("test/build/tree-shaking_side_effect", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_class() {
        let compiler = setup_compiler("test/build/tree-shaking_class", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_exported() {
        let compiler = setup_compiler("test/build/tree-shaking_exported", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_default() {
        let compiler = setup_compiler("test/build/tree-shaking_export_default", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
}
