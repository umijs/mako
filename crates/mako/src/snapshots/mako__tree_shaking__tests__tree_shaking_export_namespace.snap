---
source: crates/mako/src/tree_shaking.rs
expression: content
---
const chunksIdToUrlMap = {};
chunksIdToUrlMap["./index.ts"] = `${"/"}index.js`;
chunksIdToUrlMap["./all_vendors"] = `${"/"}all_vendors-async.js`;
function createRuntime(makoModules, entryModuleId) {
    const modulesRegistry = {};
    function requireModule(moduleId) {
        const cachedModule = modulesRegistry[moduleId];
        if (cachedModule !== undefined) {
            if (cachedModule.error) {
                throw cachedModule.error;
            }
            return cachedModule.exports;
        }
        const module = {
            id: moduleId,
            exports: {}
        };
        modulesRegistry[moduleId] = module;
        try {
            const execOptions = {
                id: moduleId,
                module,
                factory: makoModules[moduleId],
                require: requireModule
            };
            requireModule.requireInterceptors.forEach((interceptor)=>{
                interceptor(execOptions);
            });
            execOptions.factory.call(execOptions.module.exports, execOptions.module, execOptions.module.exports, execOptions.require);
        } catch (e) {
            modulesRegistry[moduleId].error = e;
            throw e;
        }
        return module.exports;
    }
    requireModule.requireInterceptors = [];
    !(function() {
        let currentParents = [];
        let currentChildModule;
        const createHmrRequire = (require, moduleId)=>{
            const me = modulesRegistry[moduleId];
            if (!me) return require;
            const fn = (request)=>{
                if (me.hot.active) {
                    if (modulesRegistry[request]) {
                        const parents = modulesRegistry[request].parents;
                        if (!parents.includes(moduleId)) {
                            parents.push(moduleId);
                        }
                    } else {
                        currentParents = [
                            moduleId
                        ];
                        currentChildModule = request;
                    }
                    if (!me.children.includes(request)) {
                        me.children.push(request);
                    }
                } else {}
                return require(request);
            };
            Object.assign(fn, require);
            return fn;
        };
        const applyHotUpdate = (chunkId, update, runtime)=>{
            const { modules, removedModules } = update;
            const outdatedModules = [];
            for (const moduleId of Object.keys(modules)){
                if (!modulesRegistry[moduleId]) continue;
                if (outdatedModules.includes(moduleId)) continue;
                outdatedModules.push(moduleId);
                const queue = [
                    moduleId
                ];
                while(queue.length){
                    const item = queue.pop();
                    const module = modulesRegistry[item];
                    if (!module) continue;
                    if (module.hot._main) {
                        location.reload();
                    }
                    if (module.hot._selfAccepted) {
                        continue;
                    }
                    for (const parentModule of module.parents){
                        if (outdatedModules.includes(parentModule)) continue;
                        outdatedModules.push(parentModule);
                        queue.push(parentModule);
                    }
                }
            }
            const outdatedSelfAcceptedModules = [];
            for (const moduleId of outdatedModules){
                const module = modulesRegistry[moduleId];
                if (module.hot._selfAccepted) {
                    outdatedSelfAcceptedModules.push(module);
                }
            }
            for (const moduleId of outdatedModules){
                const module = modulesRegistry[moduleId];
                for (const handler of module.hot._disposeHandlers){
                    handler();
                }
                module.hot.active = false;
                delete modulesRegistry[moduleId];
                for (const childModule of module.children){
                    const child = modulesRegistry[childModule];
                    if (!child) continue;
                    const idx = child.parents.indexOf(moduleId);
                    if (idx !== -1) {
                        child.parents.splice(idx, 1);
                    }
                }
            }
            registerModules(modules);
            for (const module of outdatedSelfAcceptedModules){
                module.hot._requireSelf();
            }
            runtime(requireModule);
        };
        const createModuleHotObject = (moduleId, me)=>{
            const _main = currentChildModule !== moduleId;
            const hot = {
                _acceptedDependencies: {},
                _declinedDependencies: {},
                _selfAccepted: false,
                _selfDeclined: false,
                _selfInvalidated: false,
                _disposeHandlers: [],
                _requireSelf: function() {
                    currentParents = me.parents.slice();
                    currentChildModule = _main ? undefined : moduleId;
                    requireModule(moduleId);
                },
                _main,
                active: true,
                accept () {
                    this._selfAccepted = true;
                },
                dispose (callback) {
                    this._disposeHandlers.push(callback);
                },
                invalidate () {},
                check () {
                    const current_hash = requireModule.currentHash();
                    return fetch(`/${current_hash}.hot-update.json`).then((res)=>{
                        return res.json();
                    }).then((update)=>{
                        return Promise.all(update.c.map((chunk)=>{
                            let parts = chunk.split('.');
                            let l = parts.length;
                            let left = parts.slice(0, parts.length - 1).join('.');
                            let ext = parts[l - 1];
                            const hotChunkName = [
                                left,
                                current_hash,
                                'hot-update',
                                ext
                            ].join('.');
                            return new Promise((done)=>{
                                requireModule.loadScript(`/${hotChunkName}`, done);
                            });
                        }));
                    });
                },
                apply (update) {
                    return applyHotUpdate(update);
                }
            };
            currentChildModule = undefined;
            return hot;
        };
        requireModule.requireInterceptors.push((options)=>{
            const orginRequire = options.require;
            options.module.hot = createModuleHotObject(options.id, options.module);
            options.module.meta = {
                hot: options.module.hot
            };
            options.module.parents = currentParents;
            currentParents = [];
            options.module.children = [];
            options.require = createHmrRequire(options.require, options.id);
            options.require.currentHash = ()=>{
                return orginRequire._h;
            };
        });
        requireModule.applyHotUpdate = applyHotUpdate;
    })();
    !(function() {
        requireModule.chunkEnsures = {};
        requireModule.ensure = function(chunkId) {
            return Promise.all(Object.keys(requireModule.chunkEnsures).reduce(function(promises, key) {
                requireModule.chunkEnsures[key](chunkId, promises);
                return promises;
            }, []));
        };
    })();
    !(function() {
        const installedChunks = (requireModule.jsonpInstalled = {});
        requireModule.chunkEnsures.jsonp = (chunkId, promises)=>{
            let data = installedChunks[chunkId];
            if (data === 0) return;
            if (data) {
                promises.push(data[2]);
            } else {
                const promise = new Promise((resolve, reject)=>{
                    data = installedChunks[chunkId] = [
                        resolve,
                        reject
                    ];
                });
                promises.push((data[2] = promise));
                const url = chunksIdToUrlMap[chunkId];
                const error = new Error();
                const onLoadEnd = (event)=>{
                    data = installedChunks[chunkId];
                    if (data !== 0) installedChunks[chunkId] = undefined;
                    if (data) {
                        const errorType = event?.type;
                        const src = event?.target?.src;
                        error.message = `Loading chunk ${chunkId} failed. (${errorType} : ${src})`;
                        error.name = 'ChunkLoadError';
                        error.type = errorType;
                        data[1](error);
                    }
                };
                requireModule.loadScript(url, onLoadEnd, `chunk-${chunkId}`);
                return promise;
            }
        };
    })();
    !(function() {
        const inProgress = {};
        requireModule.loadScript = (url, done, key)=>{
            if (inProgress[url]) {
                return inProgress[url].push(done);
            }
            const script = document.createElement('script');
            script.timeout = 120;
            script.src = url;
            inProgress[url] = [
                done
            ];
            const onLoadEnd = (prev, event)=>{
                clearTimeout(timeout);
                const doneFns = inProgress[url];
                delete inProgress[url];
                script.parentNode?.removeChild(script);
                if (doneFns) {
                    doneFns.forEach(function(fn) {
                        return fn(event);
                    });
                }
                if (prev) return prev(event);
            };
            const timeout = setTimeout(onLoadEnd.bind(null, undefined, {
                type: 'timeout',
                target: script
            }), 120000);
            script.onerror = onLoadEnd.bind(null, script.onerror);
            script.onload = onLoadEnd.bind(null, script.onload);
            document.head.appendChild(script);
        };
    })();
    const jsonpCallback = (data)=>{
        const installedChunks = requireModule.jsonpInstalled;
        const chunkIds = data[0];
        const modules = data[1];
        if (chunkIds.some((id)=>installedChunks[id] !== 0)) {
            registerModules(modules);
        }
        for (const id of chunkIds){
            if (installedChunks[id]) {
                installedChunks[id][0]();
            }
            installedChunks[id] = 0;
        }
    };
    const registerModules = (modules)=>{
        for(const id in modules){
            makoModules[id] = modules[id];
        }
    };
    requireModule._h = '8607444061540746814';
    requireModule.currentHash = ()=>{
        return requireModule._h;
    };
    Promise.all([
        requireModule.ensure("./all_vendors")
    ]).then(()=>{
        requireModule(entryModuleId);
    });
    return {
        requireModule,
        _modulesRegistry: modulesRegistry,
        _jsonpCallback: jsonpCallback,
        _makoModuleHotUpdate: requireModule.applyHotUpdate
    };
}
const runtime = createRuntime({
    "./1.ts": function(module, exports, require) {
        "use strict";
        Object.defineProperty(exports, "__esModule", {
            value: true
        });
        function _export(target, all) {
            for(var name in all)Object.defineProperty(target, name, {
                enumerable: true,
                get: all[name]
            });
        }
        _export(exports, {
            foo: function() {
                return foo;
            },
            bar: function() {
                return bar;
            }
        });
        const foo = 1;
        const bar = 1;
    },
    "./index.ts": function(module, exports, require) {
        "use strict";
        Object.defineProperty(exports, "__esModule", {
            value: true
        });
        var _interop_require_wildcard = require("../../../../../node_modules/.pnpm/@swc+helpers@0.5.1/node_modules/@swc/helpers/esm/_interop_require_wildcard.js");
        var _1 = _interop_require_wildcard._(require("./1.ts"));
        console.log(_1.foo);
    }
}, './index.ts');
globalThis.jsonpCallback = runtime._jsonpCallback;
globalThis.modulesRegistry = runtime._modulesRegistry;
globalThis.makoModuleHotUpdate = runtime._makoModuleHotUpdate;

//# sourceMappingURL=index.js.map
